### DISCUSSION TITLE
PlatformIo, ESP32 and LittleFS vs SPIFFS

### DISCUSSION THREAD

**Discussion by keck9939** [Bodmer/TFT_eSPI#2208] [Q&A - General]
I am trying to build the WLED project using PlatformIo targeting an ESP32 with TFT_eSPI support included. It fails to build because SPIFFS.h is not found in TFT_eSPI_ESP32.h.  The problem is, if the target is ESP32, SPIFFS.h is explicitly included in TFT_eSPI_ESP32.h. I suspect that this problem may not occur with the Arduino IDE because SPIFFS.h can be found somewhere, whereas on PlatformIo only explicitly included libraries are present. Note that WLED uses LittleFS and SPIFFS is deprecated.
I can get it to build by editing TFT_eSPI_ESP32.h, but this does not seem to be a clean solution. Is there a better way?

**Comment by Bodmer** [comment]
SPIFFS supports remains in the library for legacy sketch users and to support the included examples.

The library is written for an Arduino environment, this can be configured within PlatformIO. The library will then build without modification. You could of create a dummy header file to get around the problem or not use smooth fonts.

**Reply by keck9939** [reply]
WLED _is_ an Arduino project, although because of its dependencies the recommended build environment is PlatformIo. As such, PlatformIo is configured for Arduino. I do not believe it will build without modification without SPIFFS being present, even if it is not actually used and your reply even indicates as much. It would certainly be possible to write TFT_eSPI_ESP32.h so that it does not require SPIFFS if it is not used, for example by handling the ESP32 the same way you handle the ESP32S2.

**Reply by Bodmer** [reply]
The Arduino board package for the ESP32 includes SPIFFS so should compile fine. Provide the compiler error report.

The S2 is not handled differently as far as I can tell so it is not clear to me what you are asking.

**Reply by keck9939** [reply]
The beauty of PlatformIo is it allows you not to pull in libraries which you are not using. With the Arduino IDE, it is often completely unclear exactly which packages are being found and used. Hence, WLED does not pull in SPIFFS because it uses LittleFS not SPIFFS.
In TFT_eSPI_ESP32.h lines 133-148 it where the two processors are treated differently (the S2 will define CONFIG_IDF_TARGET_ESP32S2) and where the problem occurs.

```
// If smooth font is used then it is likely SPIFFS will be needed
#ifdef SMOOTH_FONT
  // Call up the SPIFFS (SPI FLASH Filing System) for the anti-aliased fonts
  #define FS_NO_GLOBALS
  #include <FS.h>
  #if defined(CONFIG_IDF_TARGET_ESP32)
    #include "SPIFFS.h" // ESP32 only
  #else
    #ifndef SPIFFS
      #include <LittleFS.h>
      #define SPIFFS LittleFS
    #endif
  #endif
  #define FONT_FS_AVAILABLE
#endif

```
the error reported is:
```
In file included from .pio\libdeps\esp32dev\TFT_eSPI\TFT_eSPI.h:98:0,
                 from .pio\libdeps\esp32dev\TFT_eSPI\TFT_eSPI.cpp:16:
.pio\libdeps\esp32dev\TFT_eSPI\Processors/TFT_eSPI_ESP32.h:139:38: fatal error: SPIFFS.h: No such file or directory

```
If you are not using SPIFFS you really should not need it to do the build when clearly TFT_eSPI is entirely capable of using LittleFS.

**Reply by Bodmer** [reply]
If SPIFFS.h is not included for the ESP32 then Arduino IDE users will get an error message. Including it in the sketch instead will still lead to a compile error due to the way the Arduino environment partitions things.

The ESP32 S2 core presumably loads it with FS.h as no error is generated when SPIFFS.h is not included, but SPIFFS functions are available, so there is an inconsistency in the Espressif board support package.

So to support legacy sketches it has to remain in place.

**Reply by keck9939** [reply]
I believe you are incorrect about what is really available. The `#define SPIFFS LittleFs` in the code snip-it above  means that the S2 is actually using LittleFs even if you appear to be using SPIFFS. The problem here is really they way the various processor headers in TFT_eSPI are written.
So, at this point, the answer to my original question is no.

**Reply by Bodmer** [reply]
Yes, you are right. The S2 is being handled differently from the other processors in the ESP32 family, it looks like some dev code got accidentally added. I will correct this.

SPIFFS support will remain though for the time being, in some cases it outperforms LittleFS presumably due to the way caching is handled.

Maybe the way forward is to add a setup switch for SPIFFS at some point in the future.
